### PR TITLE
feat(share/availability): persist random samples selection in availability call

### PR DESCRIPTION
**Pull request by walldiss**
This PR introduces persistence for sample selection in random sampling. It addresses the issue by storing all failed samples into the datastore, allowing them to be reloaded on the next sampling attempt. This ensures that if the availability call fails fully or partially during the last sampling attempt, the sampling retry will use the same preselected random coordinates of shares.

Provided solution is backwards compatible with previously stored empty byte slice on sampling success, allowing the change to be non-breaking for existing storage. 

Additionally, this PR includes basic refactoring to simplify concurrency logic in availability. It also ensures that errors returned by the call are aligned with the interface declaration in [availability.go](https://github.com/celestiaorg/celestia-node/blob/main/share/availability.go) enhancing code consistency and maintainability.

Resolves https://github.com/celestiaorg/celestia-node/issues/2780

### CODE REVIEW

File: share/availability/light/availability.go [hunk 1]
```diff
@@ -67,77 +66,88 @@ func (la *ShareAvailability) SharesAvailable(ctx context.Context, header *header
...
 	}
 
 	// indicate to the share.Getter that a blockservice session should be created. This
 	// functionality is optional and must be supported by the used share.Getter.
 	ctx = getters.WithSession(ctx)
 
+	result := struct {
+		sync.Mutex
+		failed []Sample
+	}{}
```
Comment (Wondertan) [inline share/availability/light/availability.go:113]: nit: this does not need to be a struct but two separate vars, which is more idiomatic

File: share/availability/light/availability.go [hunk 2]
```diff
@@ -67,77 +66,88 @@ func (la *ShareAvailability) SharesAvailable(ctx context.Context, header *header
...
-			if errors.Is(err, context.Canceled) {
-				return err
-			}
-			log.Errorw("availability validation failed", "root", dah.String(), "err", err.Error())
-			if ipldFormat.IsNotFound(err) || errors.Is(err, context.DeadlineExceeded) {
-				return share.ErrNotAvailable
-			}
-			return err
-		}
+	if errors.Is(ctx.Err(), context.Canceled) {
```
Comment (Wondertan) [inline share/availability/light/availability.go:133]: Why only context.Canceled is respected?

**Review by Wondertan — APPROVED** [review]
Very well written test

**Review by renaynay — APPROVED** [review]
Actually looks nice, minimal changes. I won't be too finicky re serialisation bc i know this will be removed / handled with shwap

File: share/availability/light/availability.go (hunk 2 above)
Comment (walldiss) [reply share/availability/light/availability.go:130]: I made it to align Availability interface according to: https://github.com/celestiaorg/celestia-node/blob/c93713f69b937a3d8bbba221e1c8026512e6d190/share/availability.go#L13

context.Canceled  is the only error that is not technically failed availability.

File: share/availability/light/sample.go [hunk 3]
```diff
@@ -66,11 +68,37 @@ func (ss *squareSampler) samples() []Sample {
...
 
-	return int(n.Int64())
+	return uint16(n.Uint64())
+}
+
+// encodeSamples encodes a slice of samples into a byte slice using little endian encoding.
+func encodeSamples(samples []Sample) []byte {
+	bs := make([]byte, 0, len(samples)*4)
+	for _, s := range samples {
+		bs = binary.LittleEndian.AppendUint16(bs, s.Row)
```
[the remark this answers is not in the document]
Comment (walldiss) [reply share/availability/light/sample.go:84]: What do you think could be better alternative?

**Review by Wondertan — APPROVED** [review]

File: share/availability/light/sample.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (renaynay) [reply share/availability/light/sample.go:84]: Was just a question why little and not big? I thought you would use big in shwap

**Review by renaynay — APPROVED** [review]

**Pull request merged**